src/covid19sim/frozen/clustering/naive.py

```diff
@@ -0,0 +1,370 @@
+    #  only correspond to the first encounter uid it receives, and it never gets updated)
+    uids: typing.Dict[np.int64, np.uint8]
+    """Timestamp-to-Unique Identifier (UID) mapping of all encounters in this cluster."""
+
+    def __init__(
+            self,
+            uid: np.uint8,
+            first_update_time: np.int64,
+            **kwargs,
+    ):
+        """Creates a naive cluster, forwarding most args to the base class."""
+        super().__init__(
+            uid=uid,
+            first_update_time=first_update_time,
+            **kwargs,
+        )
+        self.uids = {first_update_time: uid}
+
+    @staticmethod
+    def create_cluster_from_message(message: mu.GenericMessageType) -> "NaiveCluster":
+        """Creates and returns a new cluster based on a single encounter message."""
+        return NaiveCluster(
+            # app-visible stuff below
+            uid=message.uid,
+            risk_level=message.risk_level
+                if isinstance(message, mu.EncounterMessage) else message.new_risk_level,
+            first_update_time=message.encounter_time,
+            latest_update_time=message.encounter_time
+                if isinstance(message, mu.EncounterMessage) else message.update_time,
+            messages=[message] if isinstance(message, mu.EncounterMessage)
+                else mu.create_encounter_from_update_message(message),
+            # debug-only stuff below
+            _real_encounter_uids=[message._sender_uid],
+            _real_encounter_times=[message._real_encounter_time],
+            _unclustered_messages=[message],  # once added, messages here should never be removed
+        )
+
+    def _get_encounter_match_score(
+            self,
+            message: mu.EncounterMessage,
+            ticks_per_uid_roll: np.int64 = 24 * 60 * 60,  # one tick per second, one roll per day
+    ) -> typing.Tuple[int, typing.Optional[int]]:
+        """Returns the match score between the provided encounter message and this cluster.
+
+        A negative return value means the message & cluster cannot correspond to the same person. A
+        zero value means that no match can be established with certainty (e.g. because the message
+        or cluster is too old). A positive value indicates a (possibly partial) match. A high
+        positive value means the match is more likely. If the returned value is equal to
+        `message_uid_bit_count`, the match is perfect.
+
+        Returns:
+            A tuple of the best match score (an integer in `[-1,message_uid_bit_count]`) and of
+            the matched encounter index (for internal use).
+        """
+        if message.risk_level != self.risk_level:
+            # if the cluster's risk level differs from the given encounter's risk level, there is
+            # no way to match the two, as the cluster should be updated first to that risk level,
+            # or we are looking at different users entirely
+            return -1, None
```

Comment:
I'm not sure about this assumption -- older messages with different risk scores should be able to be in the same cluster, right?

Reply:
Unless clusters are day-specific

Reply:
two encounters can only be from the same person and coexist in the same cluster if they both share the same risk level; if that is not the case, then either it was two persons (and the cluster should rightfully be split), or we are in the process of updating messages and only received a partial update (in which case, we will split and partially update)

in other words, multiple encounters can coexist in a cluster only if they have all been updated to the same risk level